File: retirodelrocioapp/lib/features/housekeeping/lost_found/domain/lost_found_item.dart
import 'package:flutter/foundation.dart';

/// An item a housekeeper found while turning over a room (or a common area) —
/// logged, then eventually handed back to its owner or disposed of
/// (`GET /housekeeping/lost-found`).
@immutable
class LostFoundItem {
  const LostFoundItem({
    required this.id,
    this.roomUnitId,
    this.roomNumber,
    this.roomName,
    required this.itemDescription,
    this.notes,
    this.foundByName,
    this.foundLabel,
    required this.status,
    required this.statusLabel,
    required this.isUnclaimed,
    this.claimantName,
    this.claimantContact,
  });

  final int id;
  final int? roomUnitId;
  final String? roomNumber;
  final String? roomName;
  final String itemDescription;
  final String? notes;
  final String? foundByName;
  final String? foundLabel;

  /// unclaimed | returned | disposed.
  final String status;
  final String statusLabel;
  final bool isUnclaimed;
  final String? claimantName;
  final String? claimantContact;

  factory LostFoundItem.fromJson(Map<String, dynamic> json) => LostFoundItem(
    id: (json['id'] as num?)?.toInt() ?? 0,
    roomUnitId: (json['room_unit_id'] as num?)?.toInt(),
    roomNumber: json['room_number'] as String?,
    roomName: json['room_name'] as String?,
    itemDescription: json['item_description'] as String? ?? '',
    notes: json['notes'] as String?,
    foundByName: json['found_by_name'] as String?,
    foundLabel: json['found_label'] as String?,
    status: json['status'] as String? ?? 'unclaimed',
    statusLabel: json['status_label'] as String? ?? 'Unclaimed',
    isUnclaimed: json['is_unclaimed'] as bool? ?? true,
    claimantName: json['claimant_name'] as String?,
    claimantContact: json['claimant_contact'] as String?,
  );
}

File: retiro-del-rocio/app/Models/LostFoundItem.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

/**
 * An item a housekeeper found while turning over a room (or a common area),
 * from first logged through to being handed back or disposed of:
 * unclaimed → returned | disposed.
 */
class LostFoundItem extends Model
{
    public const UNCLAIMED = 'unclaimed';

    public const RETURNED = 'returned';

    public const DISPOSED = 'disposed';

    public const STATUSES = [self::UNCLAIMED, self::RETURNED, self::DISPOSED];

    // Matches the migration's column default. Without this, a freshly
    // `create()`d instance has a null `status` in memory until re-fetched —
    // `markReturned()`/`markDisposed()` would then see null !== UNCLAIMED and
    // silently no-op.
    protected $attributes = [
        'status' => self::UNCLAIMED,
    ];

    protected $fillable = [
        'room_unit_id', 'booking_id', 'item_description', 'notes', 'found_by', 'found_at',
        'status', 'claimant_name', 'claimant_contact', 'returned_at', 'returned_by',
    ];

    protected $casts = [
        'found_at' => 'datetime',
        'returned_at' => 'datetime',
    ];

    public function roomUnit()
    {
        return $this->belongsTo(RoomUnit::class);
    }

    public function booking()
    {
        return $this->belongsTo(Booking::class);
    }

    public function foundBy()
    {
        return $this->belongsTo(User::class, 'found_by');
    }

    public function returnedBy()
    {
        return $this->belongsTo(User::class, 'returned_by');
    }

    public function isUnclaimed(): bool
    {
        return $this->status === self::UNCLAIMED;
    }

    /** Hand the item back — unclaimed → returned. Idempotent. */
    public function markReturned(?User $officer = null, ?string $claimantName = null, ?string $claimantContact = null): bool
    {
        if (! $this->isUnclaimed()) {
            return false;
        }

        return $this->update([
            'status' => self::RETURNED,
            'claimant_name' => $claimantName ?: $this->claimant_name,
            'claimant_contact' => $claimantContact ?: $this->claimant_contact,
            'returned_at' => now(),
            'returned_by' => $officer?->id,
        ]);
    }

    /** Never claimed, discarded per policy — unclaimed → disposed. Idempotent. */
    public function markDisposed(?User $officer = null): bool
    {
        if (! $this->isUnclaimed()) {
            return false;
        }

        return $this->update([
            'status' => self::DISPOSED,
            'returned_at' => now(),
            'returned_by' => $officer?->id,
        ]);
    }

    public function statusLabel(): string
    {
        return match ($this->status) {
            self::RETURNED => 'Returned',
            self::DISPOSED => 'Disposed',
            default => 'Unclaimed',
        };
    }

    /** The payload the housekeeping tablet renders on the Lost & Found screen. */
    public function toHousekeepingArray(): array
    {
        $unit = $this->relationLoaded('roomUnit') ? $this->roomUnit : $this->roomUnit()->first();
        $room = $unit?->relationLoaded('room') ? $unit->room : $unit?->room()->first();
        $founder = $this->relationLoaded('foundBy') ? $this->foundBy : $this->foundBy()->first();

        return [
            'id' => $this->id,
            'room_unit_id' => $this->room_unit_id,
            'room_number' => $unit?->number,
            'room_name' => $room?->name,
            'item_description' => $this->item_description,
            'notes' => $this->notes,
            'found_by_name' => $founder?->name,
            'found_at' => $this->found_at?->toIso8601String(),
            'found_label' => optional($this->found_at)->diffForHumans(),
            'status' => $this->status,
            'status_label' => $this->statusLabel(),
            'is_unclaimed' => $this->isUnclaimed(),
            'claimant_name' => $this->claimant_name,
            'claimant_contact' => $this->claimant_contact,
        ];
    }
}
